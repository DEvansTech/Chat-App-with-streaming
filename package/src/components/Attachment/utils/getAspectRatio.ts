import type { Attachment } from 'stream-chat';

import type { DefaultStreamChatGenerics } from '../../../types/types';

/**
 * Returns the aspect ratio of an image attachment.
 *
 * @param image Image attachment.
 * @returns {number}
 */
export function getAspectRatio<
  StreamChatGenerics extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(image: Attachment<StreamChatGenerics>) {
  if (image.type !== 'image') {
    throw new Error('getAspectRatio() can only be called on an image attachment');
  }

  if (!image.original_width || !image.original_height) return 1;

  return image.original_width / image.original_height;
}
