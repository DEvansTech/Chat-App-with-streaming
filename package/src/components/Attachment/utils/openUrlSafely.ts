import { Linking } from 'react-native';

export const openUrlSafely = (url?: string) => {
  let finalUrl = url as string;
  const pattern = new RegExp(/^\S+:\/\//);

  if (!pattern.test(finalUrl)) {
    finalUrl = 'http://' + url;
  }

  Linking.canOpenURL(finalUrl).then((supported) => {
    if (supported) {
      Linking.openURL(finalUrl);
    } else {
      console.warn(`Don't know how to open URI: ${finalUrl}`);
    }
  });
};
