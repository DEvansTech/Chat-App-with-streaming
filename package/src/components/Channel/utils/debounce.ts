/* eslint-disable @typescript-eslint/no-explicit-any */
import type { DebounceSettings } from 'lodash';
import debounce from 'lodash/debounce';

export const heavyDebounce = <T extends (...args: any) => any>(
  callback: T,
  ms = 500,
  option: DebounceSettings = {
    leading: true,
    trailing: true,
  },
) => debounce(callback, ms, option);
