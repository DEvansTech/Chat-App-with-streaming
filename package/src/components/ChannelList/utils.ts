import type { Channel, StreamChat } from 'stream-chat';

import type { DefaultStreamChatGenerics } from '../../types/types';

type MoveParameters<
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = {
  channels: Channel<StreamChatClient>[];
  cid: string;
};

export const moveChannelUp = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  cid,
  channels = [],
}: MoveParameters<StreamChatClient>) => {
  // get channel from channels
  const index = channels.findIndex((c) => c.cid === cid);
  if (index <= 0) return channels;
  const channel = channels[index];

  // remove channel from current position and add to start
  channels.splice(index, 1);
  channels.unshift(channel);

  return [...channels];
};

type GetParameters<StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics> =
  {
    client: StreamChat<StreamChatClient>;
    id: string;
    type: string;
  };

export const getChannel = async <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>({
  client,
  id,
  type,
}: GetParameters<StreamChatClient>) => {
  const channel = client.channel(type, id);
  await channel.watch();
  return channel;
};

export const DEFAULT_QUERY_CHANNELS_LIMIT = 10;
export const MAX_QUERY_CHANNELS_LIMIT = 30;
