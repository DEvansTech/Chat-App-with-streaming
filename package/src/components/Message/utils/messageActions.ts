import type { ExtendableGenerics } from 'stream-chat';

import type { MessageContextValue } from '../../../contexts/messageContext/MessageContext';
import type { OwnCapabilitiesContextValue } from '../../../contexts/ownCapabilitiesContext/OwnCapabilitiesContext';
import type { DefaultStreamChatGenerics } from '../../../types/types';
import type { MessageActionType } from '../../MessageOverlay/MessageActionListItem';

export type MessageActionsParams<
  StreamChatClient extends ExtendableGenerics = DefaultStreamChatGenerics,
> = {
  blockUser: MessageActionType;
  copyMessage: MessageActionType;
  deleteMessage: MessageActionType;
  dismissOverlay: () => void;
  editMessage: MessageActionType;
  error: boolean | Error;
  flagMessage: MessageActionType;
  isThreadMessage: boolean;
  messageReactions: boolean;
  muteUser: MessageActionType;
  ownCapabilities: OwnCapabilitiesContextValue;
  pinMessage: MessageActionType;
  quotedReply: MessageActionType;
  retry: MessageActionType;
  threadReply: MessageActionType;
  unpinMessage: MessageActionType;
} & Pick<MessageContextValue<StreamChatClient>, 'message' | 'isMyMessage'>;

export type MessageActionsProp<
  StreamChatClient extends ExtendableGenerics = DefaultStreamChatGenerics,
> = (param: MessageActionsParams<StreamChatClient>) => MessageActionType[];

export const messageActions = <
  StreamChatClient extends ExtendableGenerics = DefaultStreamChatGenerics,
>({
  blockUser,
  copyMessage,
  deleteMessage,
  editMessage,
  error,
  flagMessage,
  isMyMessage,
  isThreadMessage,
  message,
  messageReactions,
  ownCapabilities,
  pinMessage,
  quotedReply,
  retry,
  threadReply,
  unpinMessage,
}: MessageActionsParams<StreamChatClient>) => {
  if (messageReactions) {
    return undefined;
  }

  const actions: Array<MessageActionType | null> = [];

  if (error && isMyMessage) {
    actions.push(retry);
  }

  if (ownCapabilities.quoteMessage && !isThreadMessage && !error) {
    actions.push(quotedReply);
  }

  if (ownCapabilities.sendReply && !isThreadMessage && !error) {
    actions.push(threadReply);
  }

  if (
    (isMyMessage && ownCapabilities.updateOwnMessage) ||
    (!isMyMessage && ownCapabilities.updateAnyMessage)
  ) {
    actions.push(editMessage);
  }

  if (message.text && !error) {
    actions.push(copyMessage);
  }

  if (!isMyMessage && ownCapabilities.flagMessage) {
    actions.push(flagMessage);
  }

  if (ownCapabilities.pinMessage && !message.pinned) {
    actions.push(pinMessage);
  }

  if (ownCapabilities.pinMessage && message.pinned) {
    actions.push(unpinMessage);
  }

  if (!isMyMessage && ownCapabilities.banChannelMembers) {
    actions.push(blockUser);
  }

  if (
    (isMyMessage && ownCapabilities.deleteOwnMessage) ||
    (!isMyMessage && ownCapabilities.deleteAnyMessage)
  ) {
    actions.push(deleteMessage);
  }

  return actions;
};
