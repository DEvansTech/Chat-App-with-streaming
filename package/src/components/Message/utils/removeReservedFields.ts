import type { DefaultStreamChatGenerics } from '../../../types/types';
import type { MessageType } from '../../MessageList/hooks/useMessageList';

export const removeReservedFields = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  message: MessageType<StreamChatClient>,
) => {
  const retryMessage = { ...message };
  const reserved = [
    'cid',
    'config',
    'created_at',
    'created_by',
    'id',
    'last_message_at',
    'member_count',
    'type',
    'updated_at',
  ];
  reserved.forEach((key) => {
    delete retryMessage[key];
  });

  return retryMessage;
};
