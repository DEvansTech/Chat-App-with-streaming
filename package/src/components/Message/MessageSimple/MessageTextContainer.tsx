import React from 'react';
import { StyleProp, StyleSheet, View, ViewStyle } from 'react-native';

import { renderText, RenderTextParams } from './utils/renderText';

import {
  MessageContextValue,
  useMessageContext,
} from '../../../contexts/messageContext/MessageContext';
import {
  MessagesContextValue,
  useMessagesContext,
} from '../../../contexts/messagesContext/MessagesContext';
import { useTheme } from '../../../contexts/themeContext/ThemeContext';

import type { MarkdownStyle, Theme } from '../../../contexts/themeContext/utils/theme';
import type { DefaultStreamChatGenerics } from '../../../types/types';

const styles = StyleSheet.create({
  textContainer: { maxWidth: 250, paddingHorizontal: 16 },
});

export type MessageTextProps<
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = MessageTextContainerProps<StreamChatClient> & {
  renderText: (params: RenderTextParams<StreamChatClient>) => JSX.Element | null;
  theme: { theme: Theme };
};

export type MessageTextContainerPropsWithContext<
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = Pick<
  MessageContextValue<StreamChatClient>,
  'message' | 'onLongPress' | 'onlyEmojis' | 'onPress' | 'preventPress'
> &
  Pick<MessagesContextValue<StreamChatClient>, 'markdownRules' | 'MessageText'> & {
    markdownStyles?: MarkdownStyle;
    messageOverlay?: boolean;
    messageTextNumberOfLines?: number;
    styles?: Partial<{
      textContainer: StyleProp<ViewStyle>;
    }>;
  };

const MessageTextContainerWithContext = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  props: MessageTextContainerPropsWithContext<StreamChatClient>,
) => {
  const theme = useTheme();

  const {
    markdownRules,
    markdownStyles: markdownStylesProp = {},
    message,
    messageOverlay,
    MessageText,
    messageTextNumberOfLines,
    onLongPress,
    onlyEmojis,
    onPress,
    preventPress,
    styles: stylesProp = {},
  } = props;

  const {
    theme: {
      colors,
      messageSimple: {
        content: {
          markdown,
          textContainer: { onlyEmojiMarkdown, ...textContainer },
        },
      },
    },
  } = theme;

  if (!message.text) return null;

  const markdownStyles = { ...markdown, ...markdownStylesProp };

  return (
    <View
      style={[styles.textContainer, textContainer, stylesProp.textContainer]}
      testID='message-text-container'
    >
      {MessageText ? (
        <MessageText {...props} renderText={renderText} theme={theme} />
      ) : (
        renderText<StreamChatClient>({
          colors,
          markdownRules,
          markdownStyles: {
            ...markdownStyles,
            ...(onlyEmojis ? onlyEmojiMarkdown : {}),
          },
          message,
          messageOverlay,
          messageTextNumberOfLines,
          onLongPress,
          onlyEmojis,
          onPress,
          preventPress,
        })
      )}
    </View>
  );
};

const areEqual = <StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics>(
  prevProps: MessageTextContainerPropsWithContext<StreamChatClient>,
  nextProps: MessageTextContainerPropsWithContext<StreamChatClient>,
) => {
  const { message: prevMessage, onlyEmojis: prevOnlyEmojis } = prevProps;
  const { message: nextMessage, onlyEmojis: nextOnlyEmojis } = nextProps;

  const messageTextEqual = prevMessage.text === nextMessage.text;
  if (!messageTextEqual) return false;

  const onlyEmojisEqual = prevOnlyEmojis === nextOnlyEmojis;
  if (!onlyEmojisEqual) return false;

  const mentionedUsersEqual =
    prevMessage.mentioned_users?.length === nextMessage.mentioned_users?.length &&
    (nextMessage.mentioned_users?.length === 0 ||
      (prevMessage.mentioned_users?.length &&
        nextMessage.mentioned_users?.length &&
        prevMessage.mentioned_users[0].name === nextMessage.mentioned_users[0].name));
  if (!mentionedUsersEqual) return false;

  return true;
};

const MemoizedMessageTextContainer = React.memo(
  MessageTextContainerWithContext,
  areEqual,
) as typeof MessageTextContainerWithContext;

export type MessageTextContainerProps<
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = Partial<MessageTextContainerPropsWithContext<StreamChatClient>>;

export const MessageTextContainer = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  props: MessageTextContainerProps<StreamChatClient>,
) => {
  const { message, onLongPress, onlyEmojis, onPress, preventPress } =
    useMessageContext<StreamChatClient>();
  const { markdownRules, MessageText } = useMessagesContext<StreamChatClient>();
  const { messageTextNumberOfLines } = props;

  return (
    <MemoizedMessageTextContainer
      {...{
        markdownRules,
        message,
        MessageText,
        messageTextNumberOfLines,
        onLongPress,
        onlyEmojis,
        onPress,
        preventPress,
      }}
      {...props}
    />
  );
};

MessageTextContainer.displayName = 'MessageTextContainer{messageSimple{content}}';
