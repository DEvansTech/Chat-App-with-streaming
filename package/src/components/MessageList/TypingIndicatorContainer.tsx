import React, { PropsWithChildren } from 'react';
import { StyleSheet, View } from 'react-native';

import { filterTypingUsers } from './utils/filterTypingUsers';

import { ChatContextValue, useChatContext } from '../../contexts/chatContext/ChatContext';
import { useTheme } from '../../contexts/themeContext/ThemeContext';
import { ThreadContextValue, useThreadContext } from '../../contexts/threadContext/ThreadContext';
import { TypingContextValue, useTypingContext } from '../../contexts/typingContext/TypingContext';

import type { DefaultStreamChatGenerics } from '../../types/types';

const styles = StyleSheet.create({
  container: {
    bottom: 0,
    position: 'absolute',
    width: '100%',
  },
});

type TypingIndicatorContainerPropsWithContext<
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = Pick<TypingContextValue<StreamChatClient>, 'typing'> &
  Pick<ChatContextValue<StreamChatClient>, 'client'> &
  Pick<ThreadContextValue<StreamChatClient>, 'thread'>;

const TypingIndicatorContainerWithContext = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  props: PropsWithChildren<TypingIndicatorContainerPropsWithContext<StreamChatClient>>,
) => {
  const { children, client, thread, typing } = props;

  const {
    theme: {
      messageList: { typingIndicatorContainer },
    },
  } = useTheme();
  const typingUsers = filterTypingUsers({ client, thread, typing });

  if (!typingUsers.length) {
    return null;
  }

  return (
    <View style={[styles.container, typingIndicatorContainer]} testID='typing-indicator-container'>
      {children}
    </View>
  );
};

export type TypingIndicatorContainerProps<
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
> = Partial<TypingIndicatorContainerPropsWithContext<StreamChatClient>>;

export const TypingIndicatorContainer = <
  StreamChatClient extends DefaultStreamChatGenerics = DefaultStreamChatGenerics,
>(
  props: PropsWithChildren<TypingIndicatorContainerProps<StreamChatClient>>,
) => {
  const { typing } = useTypingContext<StreamChatClient>();
  const { client } = useChatContext<StreamChatClient>();
  const { thread } = useThreadContext<StreamChatClient>();

  return <TypingIndicatorContainerWithContext {...{ client, thread, typing }} {...props} />;
};

TypingIndicatorContainer.displayName =
  'TypingIndicatorContainer{messageList{typingIndicatorContainer}}';
