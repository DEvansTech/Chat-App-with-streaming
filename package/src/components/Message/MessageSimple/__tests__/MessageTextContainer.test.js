import React from 'react';
import { Text } from 'react-native';

import { cleanup, render, waitFor } from '@testing-library/react-native';

import { OverlayProvider } from '../../../../contexts/overlayContext/OverlayProvider';
import { ThemeProvider } from '../../../../contexts/themeContext/ThemeContext';
import { defaultTheme } from '../../../../contexts/themeContext/utils/theme';
import {
  generateMessage,
  generateStaticMessage,
} from '../../../../mock-builders/generator/message';
import { generateStaticUser } from '../../../../mock-builders/generator/user';
import { Channel } from '../../../Channel/Channel';
import { Chat } from '../../../Chat/Chat';
import { MessageList } from '../../../MessageList/MessageList';
import { MessageTextContainer } from '../MessageTextContainer';

afterEach(cleanup);

describe('MessageTextContainer', () => {
  it('should render message text container', async () => {
    const staticUser = generateStaticUser(1);
    const message = generateMessage({
      user: { ...staticUser, image: undefined },
    });
    const { getByTestId, getByText, rerender, toJSON } = render(
      <OverlayProvider>
        <Chat>
          <Channel>
            <ThemeProvider style={defaultTheme}>
              <MessageList>
                <MessageTextContainer alignment='right' groupStyles={['top']} message={message} />
              </MessageList>
            </ThemeProvider>
          </Channel>
        </Chat>
      </OverlayProvider>,
    );

    await waitFor(() => {
      expect(getByTestId('message-text-container')).toBeTruthy();
      expect(getByText(message.text)).toBeTruthy();
    });

    rerender(
      <ThemeProvider style={defaultTheme}>
        <MessageList>
          <MessageTextContainer
            alignment='right'
            groupStyles={['top']}
            message={message}
            MessageText={({ message }) => <Text testID='message-text'>{message.text}</Text>}
          />
        </MessageList>
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(getByTestId('message-text-container')).toBeTruthy();
      expect(getByTestId('message-text')).toBeTruthy();
      expect(getByText(message.text)).toBeTruthy();
    });

    const staticMessage = generateStaticMessage('Hello World', {
      user: staticUser,
    });

    rerender(
      <ThemeProvider style={defaultTheme}>
        <MessageList>
          <MessageTextContainer message={staticMessage} />
        </MessageList>
      </ThemeProvider>,
    );

    await waitFor(() => {
      expect(toJSON()).toMatchSnapshot();
    });
  });
});
