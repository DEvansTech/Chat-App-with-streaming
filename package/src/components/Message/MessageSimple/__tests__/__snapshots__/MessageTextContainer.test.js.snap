// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MessageTextContainer should render message text container 1`] = `
<View
  style={
    Array [
      Object {
        "marginTop": 2,
        "paddingHorizontal": 8,
        "paddingVertical": 5,
      },
      Object {
        "alignSelf": "flex-end",
        "borderColor": "transparent",
        "borderWidth": 0,
      },
      Object {
        "backgroundColor": "#EBEBEB",
        "borderBottomLeftRadius": 16,
        "borderBottomRightRadius": 16,
        "borderTopLeftRadius": 16,
        "borderTopRightRadius": 16,
      },
      Object {},
    ]
  }
  testID="message-text-container"
>
  <View
    style={
      Array [
        Object {
          "alignSelf": "stretch",
        },
        undefined,
      ]
    }
  >
    <Text
      style={
        Object {
          "alignItems": "flex-start",
          "flexDirection": "row",
          "flexWrap": "wrap",
          "justifyContent": "flex-start",
          "marginBottom": 0,
          "marginTop": 0,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "#222222",
          }
        }
      >
        Hello World
      </Text>
    </Text>
  </View>
</View>
`;
